server/task_check_scheduler.go

```diff
@@ -177,8 +177,90 @@ func (s *TaskCheckScheduler) Register(taskType string, executor TaskCheckExecuto
 	s.executors[taskType] = executor
 }
 
+// shouldScheduleTimingTaskCheck will return whether should we schedule a timing task check for current task
+// the logic goes like the following:
+// 1. user specified the notBeforeTs field, hence a timing gate is added
+// 2. there is no running timing task check
+// 3(a). the notBeforeTs state has been altered since last check (either via new creation or patch).
+// 3(b). the notBeforeTs has NOT been altered since last check, however the last check had failed && notBeforeTs has passed now.
+// ONLY when {1 && 2 && (3.a || 3.b)} listed above would this function return true
+func (s *TaskCheckScheduler) shouldScheduleTimingTaskCheck(ctx context.Context, task *api.Task, forceSchedule bool) (bool, error) {
+	if forceSchedule {
+		return true, nil
+	}
+
+	statusList := []api.TaskCheckRunStatus{api.TaskCheckRunDone, api.TaskCheckRunFailed, api.TaskCheckRunRunning}
+	taskCheckType := api.TaskCheckGeneralEarliestAllowedTime
+	taskCheckRunFind := &api.TaskCheckRunFind{
+		TaskID:     &task.ID,
+		Type:       &taskCheckType,
+		StatusList: &statusList,
+		Latest:     true,
+	}
+	taskCheckRunList, err := s.server.TaskCheckRunService.FindTaskCheckRunList(ctx, taskCheckRunFind)
+	if err != nil {
+		return false, err
+	}
+
+	if len(taskCheckRunList) == 0 {
+		return true, nil
+	}
+
+	if taskCheckRunList[0].Status == api.TaskCheckRunRunning {
+		return false, nil
+	}
+
+	taskCheckPayload := &api.TaskCheckEarliestAllowedTimePayload{}
+	if err := json.Unmarshal([]byte(taskCheckRunList[0].Payload), taskCheckPayload); err != nil {
+		return false, err
+	}
+
+	if taskCheckPayload.NotBeforeTs != task.NotBeforeTs {
+		return true, nil
+	}
+
+	if time.Now().After(time.Unix(task.NotBeforeTs, 0)) {
+		checkResult := &api.TaskCheckRunResultPayload{}
+		if err := json.Unmarshal([]byte(taskCheckRunList[0].Result), checkResult); err != nil {
+			return false, err
+		}
+		if checkResult.ResultList[0].Status == api.TaskCheckStatusSuccess {
+			return false, nil
+		}
+		return true, nil
+	}
+
+	return false, nil
+}
+
 // ScheduleCheckIfNeeded schedules a check if needed.
 func (s *TaskCheckScheduler) ScheduleCheckIfNeeded(ctx context.Context, task *api.Task, creatorID int, skipIfAlreadyTerminated bool) (*api.Task, error) {
+	// the following block is for timing task check
+	{
+		// we only set skipIfAlreadyTerminated to false when user explicitly want to reschedule a taskCheck
+		flag, err := s.shouldScheduleTimingTaskCheck(ctx, task, !skipIfAlreadyTerminated)
```

Comment:
nit: to make it more readable, usually one would write

forceSchedule := skipIfAlreadyTerminated
flag, err := s.shouldScheduleTimingTaskCheck(ctx, task, forceSchedule)

or 

flag, err := s.shouldScheduleTimingTaskCheck(ctx, task, !skipIfAlreadyTerminated /* forceSchedule */)